File: src/main/java/com/weeth/global/auth/apple/AppleAuthService.java
package com.weeth.global.auth.apple;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import com.weeth.global.auth.apple.dto.ApplePublicKey;
import com.weeth.global.auth.apple.dto.ApplePublicKeys;
import com.weeth.global.auth.apple.dto.AppleTokenResponse;
import com.weeth.global.auth.apple.dto.AppleUserInfo;
import com.weeth.global.auth.apple.exception.AppleAuthenticationException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestClient;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.spec.RSAPublicKeySpec;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Base64;
import java.util.Date;
import java.util.Map;

@Service
@Slf4j
public class AppleAuthService {

    @Value("${auth.providers.apple.client_id}")
    private String appleClientId;

    @Value("${auth.providers.apple.team_id}")
    private String appleTeamId;

    @Value("${auth.providers.apple.key_id}")
    private String appleKeyId;

    @Value("${auth.providers.apple.redirect_uri}")
    private String redirectUri;

    @Value("${auth.providers.apple.token_uri}")
    private String tokenUri;

    @Value("${auth.providers.apple.keys_uri}")
    private String keysUri;

    @Value("${auth.providers.apple.private_key_path}")
    private String privateKeyPath;

    @Value("${auth.providers.apple.allowed_audiences}")
    private java.util.List<String> allowedAudiences;

    private final RestClient restClient = RestClient.create();

    /**
     * Authorization code로 애플 토큰 요청
     * client_secret은 JWT로 생성 (ES256 알고리즘)
     */
    public AppleTokenResponse getAppleToken(String authCode) {
        String clientSecret = generateClientSecret();

        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
        body.add("grant_type", "authorization_code");
        body.add("client_id", appleClientId);
        body.add("client_secret", clientSecret);
        body.add("code", authCode);
        body.add("redirect_uri", redirectUri);

        return restClient.post()
                .uri(tokenUri)
                .body(body)
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .retrieve()
                .body(AppleTokenResponse.class);
    }

    /**
     * ID Token 검증 및 사용자 정보 추출
     * 애플은 별도 userInfo 엔드포인트가 없고 ID Token에 정보가 포함됨
     */
    public AppleUserInfo verifyAndDecodeIdToken(String idToken) {
        try {
            // 1. ID Token의 헤더에서 kid 추출
            String[] tokenParts = idToken.split("\\.");
            String header = new String(Base64.getUrlDecoder().decode(tokenParts[0]));
            Map<String, Object> headerMap = parseJson(header);
            String kid = (String) headerMap.get("kid");

            // 2. 애플 공개키 가져오기
            ApplePublicKeys publicKeys = restClient.get()
                    .uri(keysUri)
                    .retrieve()
                    .body(ApplePublicKeys.class);

            // 3. kid와 일치하는 공개키 찾기
            ApplePublicKey matchedKey = publicKeys.keys().stream()
                    .filter(key -> key.kid().equals(kid))
                    .findFirst()
                    .orElseThrow(AppleAuthenticationException::new);

            // 4. 공개키로 ID Token 검증
            PublicKey publicKey = generatePublicKey(matchedKey);
            // JJWT 0.13.0+ uses parser() instead of parserBuilder()
            Claims claims = Jwts.parser()
                    .verifyWith(publicKey)
                    .build()
                    .parseSignedClaims(idToken)
                    .getPayload();

            // 5. Claims 검증
            validateClaims(claims);

            // 6. 사용자 정보 추출
            String appleId = claims.getSubject();
            String email = claims.get("email", String.class);
            Boolean emailVerified = claims.get("email_verified", Boolean.class);

            return AppleUserInfo.builder()
                    .appleId(appleId)
                    .email(email)
                    .emailVerified(emailVerified != null ? emailVerified : false)
                    .build();

        } catch (Exception e) {
            log.error("애플 ID Token 검증 실패", e);
            throw new AppleAuthenticationException();
        }
    }

    /**
     * 애플 로그인용 client_secret 생성
     * ES256 알고리즘으로 JWT 생성 (p8 키 파일 사용)
     */
    private String generateClientSecret() {
        try (InputStream inputStream = getInputStream(privateKeyPath)) {
            // p8 파일에서 Private Key 읽기
            String privateKeyContent = new String(inputStream.readAllBytes(), StandardCharsets.UTF_8);

            // PEM 형식의 헤더/푸터 제거
            privateKeyContent = privateKeyContent
                    .replace("-----BEGIN PRIVATE KEY-----", "")
                    .replace("-----END PRIVATE KEY-----", "")
                    .replaceAll("\\s", "");

            // Private Key 객체 생성
            byte[] keyBytes = Base64.getDecoder().decode(privateKeyContent);
            KeyFactory keyFactory = KeyFactory.getInstance("EC");
            PrivateKey privateKey = keyFactory.generatePrivate(
                    new java.security.spec.PKCS8EncodedKeySpec(keyBytes)
            );

            // JWT 생성
            LocalDateTime now = LocalDateTime.now();
            Date issuedAt = Date.from(now.atZone(ZoneId.systemDefault()).toInstant());
            Date expiration = Date.from(now.plusMonths(5).atZone(ZoneId.systemDefault()).toInstant());

            return Jwts.builder()
                    .setHeaderParam("kid", appleKeyId)
                    .setHeaderParam("alg", "ES256")
                    .setIssuer(appleTeamId)
                    .setIssuedAt(issuedAt)
                    .setExpiration(expiration)
                    .setAudience("https://appleid.apple.com")
                    .setSubject(appleClientId)
                    .signWith(privateKey, SignatureAlgorithm.ES256)
                    .compact();

        } catch (Exception e) {
            log.error("애플 Client Secret 생성 실패", e);
            throw new AppleAuthenticationException();
        }
    }

    /**
     * 파일 경로에서 InputStream 가져오기
     * 절대 경로면 파일 시스템에서, 상대 경로면 classpath에서 읽음
     */
    private InputStream getInputStream(String path) throws IOException {
        // 절대 경로인 경우 파일 시스템에서 읽기
        if (path.startsWith("/") || path.matches("^[A-Za-z]:.*")) {
            return new FileInputStream(path);
        }
        // 상대 경로는 classpath에서 읽기
        return new ClassPathResource(path).getInputStream();
    }

    /**
     * 애플 공개키로부터 PublicKey 객체 생성
     */
    private PublicKey generatePublicKey(ApplePublicKey applePublicKey) {
        try {
            byte[] nBytes = Base64.getUrlDecoder().decode(applePublicKey.n());
            byte[] eBytes = Base64.getUrlDecoder().decode(applePublicKey.e());

            BigInteger n = new BigInteger(1, nBytes);
            BigInteger e = new BigInteger(1, eBytes);

            RSAPublicKeySpec publicKeySpec = new RSAPublicKeySpec(n, e);
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");

            return keyFactory.generatePublic(publicKeySpec);
        } catch (Exception ex) {
            log.error("애플 공개키 생성 실패", ex);
            throw new AppleAuthenticationException();
        }
    }

    /**
     * ID Token의 Claims 검증
     */
    private void validateClaims(Claims claims) {
        String iss = claims.getIssuer();
        // JJWT 0.13.0+ returns Set<String> for getAudience()
        var audSet = claims.getAudience();
        String aud = audSet.iterator().hasNext() ? audSet.iterator().next() : null;

        if (!iss.equals("https://appleid.apple.com")) {
            throw new RuntimeException("유효하지 않은 발급자(issuer)입니다.");
        }

        // 허용된 audience 목록에 포함되어 있는지 확인 (웹 + Leenk 앱)
        if (aud == null || !allowedAudiences.contains(aud)) {
            log.error("유효하지 않은 audience: {}. 허용된 목록: {}", aud, allowedAudiences);
            throw new RuntimeException("유효하지 않은 수신자(audience)입니다.");
        }

        Date expiration = claims.getExpiration();
        if (expiration.before(new Date())) {
            throw new RuntimeException("만료된 ID Token입니다.");
        }
    }

    /**
     * JSON 문자열을 Map으로 파싱
     */
    @SuppressWarnings("unchecked")
    private Map<String, Object> parseJson(String json) {
        try {
            com.fasterxml.jackson.databind.ObjectMapper objectMapper = new com.fasterxml.jackson.databind.ObjectMapper();
            return objectMapper.readValue(json, Map.class);
        } catch (Exception e) {
            throw new RuntimeException("JSON 파싱 실패");
        }
    }
}

File: src/main/kotlin/com/weeth/global/auth/apple/AppleAuthService.kt
package com.weeth.global.auth.apple

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ObjectNode
import com.weeth.global.auth.apple.dto.ApplePublicKey
import com.weeth.global.auth.apple.dto.ApplePublicKeys
import com.weeth.global.auth.apple.dto.AppleTokenResponse
import com.weeth.global.auth.apple.dto.AppleUserInfo
import com.weeth.global.auth.apple.exception.AppleAuthenticationException
import com.weeth.global.config.properties.OAuthProperties
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import org.slf4j.LoggerFactory
import org.springframework.core.io.ClassPathResource
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.client.RestClient
import org.springframework.web.client.body
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import java.math.BigInteger
import java.nio.charset.StandardCharsets
import java.security.KeyFactory
import java.security.PrivateKey
import java.security.PublicKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.RSAPublicKeySpec
import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.util.Base64
import java.util.Date

@Service
class AppleAuthService(
    oAuthProperties: OAuthProperties,
    restClientBuilder: RestClient.Builder,
    private val objectMapper: ObjectMapper,
    private val clock: Clock = Clock.systemUTC(),
) {
    private data class CachedKeys(
        val keys: ApplePublicKeys,
        val expiresAt: Instant,
    )

    private val log = LoggerFactory.getLogger(javaClass)

    private val appleProperties = oAuthProperties.apple
    private val restClient = restClientBuilder.build()
    private val publicKeysTtl: Duration = Duration.ofHours(1)

    @Volatile private var cached: CachedKeys? = null
    private val privateKey: PrivateKey by lazy { loadPrivateKey() }

    fun getAppleToken(authCode: String): AppleTokenResponse {
        val clientSecret = generateClientSecret()

        val body =
            LinkedMultiValueMap<String, String>().apply {
                add("grant_type", "authorization_code")
                add("client_id", appleProperties.clientId)
                add("client_secret", clientSecret)
                add("code", authCode)
                add("redirect_uri", appleProperties.redirectUri)
            }

        return requireNotNull(
            restClient
                .post()
                .uri(appleProperties.tokenUri)
                .body(body)
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .retrieve()
                .body<AppleTokenResponse>(),
        )
    }

    fun verifyAndDecodeIdToken(idToken: String): AppleUserInfo {
        try {
            val tokenParts = idToken.split(".")
            if (tokenParts.size < 2) {
                throw AppleAuthenticationException()
            }
            val header = decodeBase64Url(tokenParts[0])
            val headerJson = parseJson(header)
            val kid = headerJson["kid"]?.asText()?.takeIf { it.isNotBlank() } ?: throw AppleAuthenticationException()
            val alg = headerJson["alg"]?.asText()
            if (alg != "RS256") {
                throw AppleAuthenticationException()
            }

            val publicKeys = getApplePublicKeys()

            val matchedKey =
                publicKeys.keys
                    .firstOrNull { key -> key.kid == kid }
                    ?: throw AppleAuthenticationException()

            val publicKey = generatePublicKey(matchedKey)
            val claims =
                Jwts
                    .parser()
                    .verifyWith(publicKey)
                    .build()
                    .parseSignedClaims(idToken)
                    .payload

            validateClaims(claims)

            val appleId = claims.subject
            val email = claims.get("email", String::class.java)
            val emailVerified = parseEmailVerified(claims["email_verified"])
            val name = claims.get("name", String::class.java)

            return AppleUserInfo(
                appleId = appleId,
                email = email,
                emailVerified = emailVerified,
                name = name,
            )
        } catch (e: AppleAuthenticationException) {
            throw e
        } catch (e: Exception) {
            log.error("애플 ID Token 검증 실패", e)
            throw AppleAuthenticationException()
        }
    }

    private fun generateClientSecret(): String {
        try {
            val now = Instant.now(clock)
            val expiration = now.plus(Duration.ofDays(150)) // Apple limit is <= 6 months.

            return Jwts
                .builder()
                .header()
                .keyId(appleProperties.keyId)
                .and()
                .issuer(appleProperties.teamId)
                .issuedAt(Date.from(now))
                .expiration(Date.from(expiration))
                .audience()
                .add("https://appleid.apple.com")
                .and()
                .subject(appleProperties.clientId)
                .signWith(privateKey, Jwts.SIG.ES256)
                .compact()
        } catch (e: Exception) {
            log.error("애플 Client Secret 생성 실패", e)
            throw AppleAuthenticationException()
        }
    }

    private fun loadPrivateKey(): PrivateKey =
        try {
            getInputStream(appleProperties.privateKeyPath).use { inputStream ->
                var privateKeyContent = String(inputStream.readAllBytes(), StandardCharsets.UTF_8)
                privateKeyContent =
                    privateKeyContent
                        .replace("-----BEGIN PRIVATE KEY-----", "")
                        .replace("-----END PRIVATE KEY-----", "")
                        .replace("\\s".toRegex(), "")

                val keyBytes = Base64.getDecoder().decode(privateKeyContent)
                val keyFactory = KeyFactory.getInstance("EC")
                keyFactory.generatePrivate(PKCS8EncodedKeySpec(keyBytes))
            }
        } catch (e: Exception) {
            log.error("애플 개인키 로드 실패", e)
            throw AppleAuthenticationException()
        }

    @Throws(IOException::class)
    private fun getInputStream(path: String): InputStream =
        if (path.startsWith("/") || path.matches(Regex("^[A-Za-z]:.*"))) {
            FileInputStream(path)
        } else {
            ClassPathResource(path).inputStream
        }

    private fun generatePublicKey(applePublicKey: ApplePublicKey): PublicKey =
        try {
            val nBytes = Base64.getUrlDecoder().decode(applePublicKey.n)
            val eBytes = Base64.getUrlDecoder().decode(applePublicKey.e)

            val n = BigInteger(1, nBytes)
            val e = BigInteger(1, eBytes)

            val publicKeySpec = RSAPublicKeySpec(n, e)
            val keyFactory = KeyFactory.getInstance("RSA")

            keyFactory.generatePublic(publicKeySpec)
        } catch (ex: Exception) {
            log.error("애플 공개키 생성 실패", ex)
            throw AppleAuthenticationException()
        }

    private fun validateClaims(claims: Claims) {
        val iss = claims.issuer
        val audiences = claims.audience
        val expiration = claims.expiration
        val now = Date.from(Instant.now(clock))

        when {
            iss != "https://appleid.apple.com" -> {
                log.warn("유효하지 않은 발급자: {}", iss)
                throw AppleAuthenticationException()
            }

            audiences.isEmpty() || !audiences.contains(appleProperties.clientId) -> {
                log.warn("유효하지 않은 audience: {}. 기대값: {}", audiences, appleProperties.clientId)
                throw AppleAuthenticationException()
            }

            expiration.before(now) -> {
                log.warn("만료된 ID Token")
                throw AppleAuthenticationException()
            }

            claims.subject.isNullOrBlank() -> {
                log.warn("유효하지 않은 subject")
                throw AppleAuthenticationException()
            }
        }
    }

    private fun getApplePublicKeys(): ApplePublicKeys {
        val now = Instant.now(clock)
        cached?.let {
            if (now.isBefore(it.expiresAt)) {
                return it.keys
            }
        }

        val fetched =
            requireNotNull(
                restClient
                    .get()
                    .uri(appleProperties.keysUri)
                    .retrieve()
                    .body<ApplePublicKeys>(),
            )

        cached = CachedKeys(fetched, now.plus(publicKeysTtl))
        return fetched
    }

    private fun parseJson(json: String): ObjectNode =
        try {
            objectMapper.readTree(json) as? ObjectNode ?: throw AppleAuthenticationException()
        } catch (e: Exception) {
            throw AppleAuthenticationException()
        }

    private fun decodeBase64Url(value: String): String = String(Base64.getUrlDecoder().decode(value), StandardCharsets.UTF_8)

    private fun parseEmailVerified(raw: Any?): Boolean =
        when (raw) {
            is Boolean -> raw
            is String -> raw.toBooleanStrictOrNull() ?: false
            else -> false
        }
}
